File: Algorithms/C++/814-Binary_Tree_Pruning.cpp
// Runtime: 4 ms, faster than 59.48% of C++ online submissions for Binary Tree Pruning.
// Memory Usage: 8 MB, less than 100.00% of C++ online submissions for Binary Tree Pruning.
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* pruneTree(TreeNode* root) {
        if (root == nullptr)
            return nullptr;
        root->left  = this->pruneTree(root->left);
        root->right = this->pruneTree(root->right);
        if (root->val == 0 && root->left == nullptr && root->right == nullptr)
            root = nullptr;
        return root;
    }
};

File: Algorithms/Python3.x/814-Binary_Tree_Pruning.py
# Runtime: 60 ms, faster than 6.48% of Python3 online submissions for Binary Tree Pruning.
# Memory Usage: 13.8 MB, less than 11.11% of Python3 online submissions for Binary Tree Pruning.
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pruneTree(self, root: TreeNode) -> TreeNode:
        if root is None:
            return None
        root.left  = self.pruneTree(root.left)
        root.right = self.pruneTree(root.right)
        if root.val == 0 and root.left is None and root.right is None:
            root = None
        return root
